server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FilterByFilterAggregator.java

```diff
@@ -109,6 +114,21 @@ final void add(QueryToFilterAdapter<?> filter) throws IOException {
                 valid = false;
                 return;
             }
+            if (filters.size() == 1) {
+                /*
+                 * When we add the second filter we check if there are any _doc_count
+                 * fields and bail out of filter-by filter mode if there are. _doc_count
+                 * fields are expensive to decode and the overhead of iterating per
+                 * filter causes us to decode doc counts over and over again.
+                 */
+                Term term = new Term(DocCountFieldMapper.NAME, DocCountFieldMapper.NAME);
+                for (LeafReaderContext c : context.searcher().getLeafContexts()) {
```

Comment:
I wonder if this is worth moving up to the `AggregationContext`.  I feel like a `hasDocCountField()` method might be useful beyond this, and I suspect it is intuitive to look for such a method on `AggregationContext`, but less so to look for it here.  Easy enough to do in a follow up if we want to though.

Reply:
Makes sense to me.